### PR TITLE
use bindgen to generate ffi

### LINKED ISSUE #128: investigate servo/rust-bindgen usage
It is error-prone to hand-maintain a mapping to the underlying rocksdb library. We may want to switch over to using https://github.com/servo/rust-bindgen (just against the rocksdb c API initially) in `librocksdb-sys`'s `build.rs`.

### PR DESCRIPTION
**Pull request by benesch**
Closes #128.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -119,29 +119,23 @@ pub struct BlockBasedOptions {
...
+/// opts.set_max_write_buffer_number(32);
+/// opts.set_write_buffer_size(536870912);
+/// opts.set_target_file_size_base(1073741824);
+/// opts.set_min_write_buffer_number_to_merge(4);
+/// opts.set_level_zero_stop_writes_trigger(2000);
+/// opts.set_level_zero_slowdown_writes_trigger(0);
+/// opts.set_compaction_style(DBCompactionStyle::Universal);
+/// opts.set_max_background_compactions(4);
+/// opts.set_max_background_flushes(4);
+/// opts.set_disable_auto_compactions(true);
```
Comment (spacejam) [inline src/lib.rs]: Can you leave the part about `badly_tuned_for_somebody_elses_disk`? Far too often, people look up a tunable, and blindly copy what they find in docs somewhere. I would like to encourage people to measure reasonable values for their own workloads and available hardware.

File: librocksdb-sys/src/lib.rs [hunk 2]
```diff
@@ -11,1024 +11,11 @@
...
-//
-
-//! Raw bindings for RocksDB.
-//!
-//! This is simply a raw interface to the RocksDB C API. It is intended to underpin a higher-level library rather than for direct use.
 
-#![allow(dead_code, non_camel_case_types, non_upper_case_globals, non_snake_case)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+#![allow(non_upper_case_globals)]
```
Comment (spacejam) [inline librocksdb-sys/src/lib.rs]: I think adding `#![allow(dead_code)]` here may address the current build error, as it's generating >4mb of unused variable warnings, causing travis to abort.

Comment (spacejam) [reply librocksdb-sys/src/lib.rs]: actually, this is being hit due to warnings emitted during the C++ compilation, and the issue is also happining on master right now, so I think the CI environment has changed. I'll try to address this on master and I'll re-trigger these tests!

Comment (benesch) [reply librocksdb-sys/src/lib.rs]: Awesome, thanks @spacejam!

**Comment by spacejam** [comment]
I've fixed the issue with travis aborting the build after 4mb of warning messages, and now the build is failing on this:
```
warning: unused macro definition

  --> librocksdb-sys/tests/ffi.rs:41:1

   |

41 | / macro_rules! cstr {

42 | |     ($($arg:tt)*) => (const_cstr!($($arg)*));

43 | | }

   | |_^

   |

   = note: #[warn(unused_macros)] on by default

error[E0308]: mismatched types

   --> librocksdb-sys/tests/ffi.rs:432:50

    |

432 |         rocksdb_options_set_compression(options, no_compression);

    |                                                  ^^^^^^^^^^^^^^ expected i32, found enum `ffi::_bindgen_ty_3`

    |

    = note: expected type `i32`

               found type `ffi::_bindgen_ty_3`

error[E0308]: mismatched types

   --> librocksdb-sys/tests/ffi.rs:902:47

    |

902 |                                               cf_names.as_ptr(),

    |                                               ^^^^^^^^^^^^^^^^^ types differ in mutability

    |

    = note: expected type `*mut *const i8`

               found type `*const *const i8`

error[E0308]: mismatched types

   --> librocksdb-sys/tests/ffi.rs:903:47

    |

903 |                                               cf_opts.as_ptr(),

    |                                               ^^^^^^^^^^^^^^^^ types differ in mutability

    |

    = note: expected type `*mut *const ffi::rocksdb_options_t`

               found type `*const *const ffi::rocksdb_options_t`

error: aborting due to 3 previous errors
```

**Comment by tschottdorf** [comment]
Ah, I believe @benesch  knows how to fix those. So close!

**Pull request merged**